pip/pip-323.md

```diff
@@ -0,0 +1,167 @@
+
+For each topic which has exceeded its quota, if the policy chosen is eviction, then the process it performed
+synchronously. This process consumes I/O, as it needs read messages (using skip) to know where to stop acknowledging
+messages.
+
+
+# Motivation
+
+Users which have defined backlog quota based on time, have no means today to monitor the backlog quota usage, 
+time-wise, to know whether the topic backlog is close to its time limit or even passed it.
+
+If it has passed it, the user has no means to know if it happened, when and how many times.
+
+
+# Goals
+
+## In Scope
+- Allow the user to know the backlog quota usage for time-based quota, per topic
+- Allow the user to know how many times backlog eviction happened, and for which backlog quota type
+
+## Out of Scope
+
+None
+
+
+# High Level Design
+
+We'll use the existing backlog monitoring process running in intervals. For each topic, the subscription with 
+the oldest unacknowledged message is retrieved, to calculate the topic backlog age. At that point, we will
+cache the following for the oldest unacknowledged message:
+* Subscription name 
+* Message position
+* Message publish timestamp
+
+That cache will allow us to add a metric exposing the topic backlog age - `pulsar_storage_backlog_age_seconds`, 
+which will be both consistent (same ones used for deciding on backlog eviction) and cheap to retrieve 
+(no additional I/O involved). 
+Coupled with the existing `pulsar_storage_backlog_quota_limit_time` metric, the user can use both to divide and
+get the usage of the quota (both are in seconds units).
+
+We will add the subscription name containing the oldest unacknowledged message to the Admin API
+topic stats endpoints (`{tenant}/{namespace}/{topic}/stats` and `{tenant}/{namespace}/{topic}/partitioned-stats`),
+allowing the user a complete workflow: alert using metrics when topic backlog is about to be exceeded, then
+query topic stats for that topic to retrieve the subscription name which contains the oldest message.
+For completeness, we will also add the backlog quota limits, both age and size, and the age of oldest 
+unacknowledged message.
+
+We will add a metric allowing the user to know how many times the usage exceeded the quota, both for time or size -
+`pulsar_storage_backlog_quota_exceeded_evictions_total`, where the `quota_type` label will be either `time` or 
+`size`. Monitoring that counter over time will allow the user to know when a topic backlog exceeded its quota,
+and if backlog eviction was chosen as action, then it happened, and how many times. 
+
+Some users may want the backlog quota check to happen more frequently, and as a consequence, the backlog age 
+metric more frequently updated. They can modify `backlogQuotaCheckIntervalInSeconds` configuration key, but without
+knowing how long this check takes, it will be hard for them. Hence, we will add the metric
+`pulsar_storage_backlog_quota_check_duration_seconds` which will be of histogram type.
```

Comment:
I was primarily talking about running `monitorBacklogQuota` in higher frequency resulting in metrics data being more fresh. 
I was thinking that in the PR I might change this to FixedDelay mainly to prevent backlogging queue of `monitorBacklogQuota` tasks which seems not reasonable to do so.